### ISSUE TITLE
Incorrect type cast when TypeGuard used with TypeVar

### ISSUE THREAD

**Issue by a-shchupakov** [microsoft/pyright#5400]
**Describe the bug**
I am using `TypeGuard` on `TypeVar`
After applying type guard on `Literal["some_value"]`, object is casted to `str`, instead of `Literal["some_value"]`

**Code or Screenshots**
```python
import typing


TValue = typing.TypeVar("TValue")
Nullable = typing.Union[TValue, None]


def is_value_set(value: Nullable[TValue]) -> typing.TypeGuard[TValue]:
    return value is not None


StatusType = typing.Literal["open", "closed"]


def print_status(status_to_print: StatusType) -> None:
    print(status_to_print)


status: Nullable[StatusType] = "open"


if is_value_set(status):
    print_status(status)  # Argument of type "str" cannot be assigned to parameter "status_to_print" of type "StatusType" in function "print_status"

```

**Expected behavior**
No errors raised

**Actual behavior**
```
Argument of type "str" cannot be assigned to parameter "status_to_print" of type "StatusType" in function "print_status"
    Type "str" cannot be assigned to type "StatusType"
      "str" cannot be assigned to type "Literal['open']"
      "str" cannot be assigned to type "Literal['closed']" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Using Pyright Command-Line:
```
pyright --version    
pyright 1.1.316
```

**Comment by erictraut** [comment]
Pyright is working as intended here.

Pyright's constraint solver (the logic that is responsible for "solving" type variables when calling a generic function) does not retain literal types unless they are required to satisfy the constraints. For more details, search for "constraint solver" in pyright's documentation.

You can force pyright to retain literal values by adding a TypeVar constraint such as a "bound" that includes a literal type.

```python
TValue = typing.TypeVar("TValue", bound="StatusType")
```

or

```python
TValue = typing.TypeVar("TValue", bound=typing.LiteralString)
```
